ichnaea/data/schema.py

```diff
@@ -0,0 +1,376 @@
+    MAX_HEADING,
+    MAX_LAT,
+    MAX_SPEED,
+    MIN_ALTITUDE,
+    MIN_LAT,
+    VALID_WIFI_REGEX,
+)
+from ichnaea.models import (
+    MAX_RADIO_TYPE,
+    MIN_RADIO_TYPE,
+    RADIO_TYPE,
+)
+from ichnaea import util
+
+
+# Utility functions
+
+def normalized_time(time):
+    """
+    Takes a string representation of a time value, validates and parses
+    it and returns a JSON-friendly string representation of the normalized
+    time.
+    """
+    now = util.utcnow()
+    if not time:
+        time = None
+
+    try:
+        time = iso8601.parse_date(time)
+    except (iso8601.ParseError, TypeError):
+        time = now
+    else:
+        # don't accept future time values or
+        # time values more than 60 days in the past
+        min_time = now - timedelta(days=60)
+        if time > now or time < min_time:
+            time = now
+    # cut down the time to a monthly resolution
+    time = time.date().replace(day=1)
+    return encode_datetime(time)
+
+
+def normalized_wifi_key(key):
+    """
+    Remove ':' '-' ',' from a wifi key.
+    """
+    if ":" in key or "-" in key or "." in key:
+        key = key.replace(":", "").replace("-", "").replace(".", "")
+    return key.lower()
+
+
+def valid_wifi_pattern(key):
+    """
+    Return True if a wifi key matches our valid regex,
+    our invalid regex, and has length 12.
+    """
+    return INVALID_WIFI_REGEX.match(key) and \
+        VALID_WIFI_REGEX.match(key) and len(key) == 12
+
+
+# Custom Types
+
+class DateTimeFromString(DateTime):
+    """
+    A DateTimeFromString will return a datetime object
+    from either a datetime object or a string.
+    """
+
+    def deserialize(self, schema, cstruct):
+        if type(cstruct) == datetime.datetime:
+            return cstruct
+        return super(DateTimeFromString, self).deserialize(schema, cstruct)
+
+
+class DateTimeToString(String):
+    """
+    A DateTimeToString will return a string respresentation of a date
+    from either a datetime object or a string.
+    """
+
+    def deserialize(self, schema, cstruct):
+        if type(cstruct) == datetime.datetime:
+            cstruct = cstruct.strftime('%Y-%m-%d')
+        return super(DateTimeToString, self).deserialize(schema, cstruct)
+
+
+# Custom Nodes
+
+class DefaultNode(SchemaNode):
+    """
+    A DefaultNode will use its ``missing`` value
+    if it fails to validate during deserialization.
+    """
+
+    def deserialize(self, cstruct):
+        try:
+            return super(DefaultNode, self).deserialize(cstruct)
+        except Invalid:
+            return self.missing
+
+
+class WifiKeyNode(SchemaNode):
+    """
+    A node containing a valid wifi key.
+    ex: 01005e901000
+    """
+
+    def preparer(self, cstruct):
+        return normalized_wifi_key(cstruct)
+
+    def validator(self, node, cstruct):
+        if not valid_wifi_pattern(cstruct):
+            raise Invalid(node, 'Invalid wifi key')
+
+
+class ReportNode(SchemaNode):
```

Comment:
I'd prefer ReportIdNode, a "report" is a full schema with lots of nodes in it, not just a single node.